src/main/java/bio/terra/app/controller/SearchApiController.java

```diff
@@ -15,30 +25,72 @@
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestParam;
 
+import javax.servlet.http.HttpServletRequest;
 import javax.validation.Valid;
+import java.util.Optional;
+import java.util.UUID;
 
 @Controller
 @Api(tags = {"search"})
 @ConditionalOnProperty(name = "features.search.api", havingValue = "enabled")
 public class SearchApiController implements SearchApi {
 
+    private final ObjectMapper objectMapper;
+    private final HttpServletRequest request;
+    private final IamService iamService;
+    private final AuthenticatedUserRequestFactory authenticatedUserRequestFactory;
+    private final SearchService searchService;
+    private final SnapshotService snapshotService;
+
     @Autowired
-    public SearchApiController() {
-        // do nothing
+    public SearchApiController(
+        ObjectMapper objectMapper,
+        HttpServletRequest request,
+        IamService iamService,
+        AuthenticatedUserRequestFactory authenticatedUserRequestFactory,
+        SearchService searchService,
+        SnapshotService snapshotService
+    ) {
+        this.objectMapper = objectMapper;
+        this.request = request;
+        this.iamService = iamService;
+        this.authenticatedUserRequestFactory = authenticatedUserRequestFactory;
+        this.searchService = searchService;
+        this.snapshotService = snapshotService;
+    }
+
+    @Override
+    public Optional<ObjectMapper> getObjectMapper() {
+        return Optional.of(objectMapper);
+    }
+
+    @Override
+    public Optional<HttpServletRequest> getRequest() {
+        return Optional.ofNullable(request);
+    }
+
+    private AuthenticatedUserRequest getAuthenticatedInfo() {
+        return authenticatedUserRequestFactory.from(request);
     }
 
     @Override
     public ResponseEntity<SearchIndexModel> createSearchIndex(
         @PathVariable("id") String id,
-        @Valid @RequestBody SearchIndexRequest body
+        @Valid @RequestBody SearchIndexRequest searchIndexRequest
     ) {
-        SearchIndexModel searchIndexModel = new SearchIndexModel();
-        return new ResponseEntity<>(searchIndexModel, HttpStatus.OK);
+        iamService.verifyAuthorization(getAuthenticatedInfo(), IamResourceType.DATASET, id, IamAction.LINK_SNAPSHOT);
+        Snapshot snapshot = snapshotService.retrieve(UUID.fromString(id));
+        try {
+            SearchIndexModel searchIndexModel = searchService.indexSnapshot(snapshot, searchIndexRequest);
+            return new ResponseEntity<>(searchIndexModel, HttpStatus.OK);
```

Comment:
Not a huge change but I think the `ok()` factory is a bit simpler than using `new`
```suggestion
            return ResponseEntity.ok(searchService.indexSnapshot(snapshot, searchIndexRequest));
```

Reply:
All our other controllers use the other formulation. Should I switch it if it is inconsistent with our other controllers?